File: LeetCode Solutions/April LeetCoding Challenge/Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        slow, fast, previous = head, head, None
        while(fast and fast.next):
            slow, fast = slow.next, fast.next.next
        previous, slow, previous.next = slow, slow.next, None
        while(slow):
            temp = slow.next
            slow.next, previous, slow = previous, slow, temp
        fast, slow = head, previous
        while(slow):
            if(fast.val != slow.val):
                return False
            fast, slow = fast.next, slow.next
        return True

File: LeetCode Solutions/August 2022 LeetCoding Challenge/Palindrome Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


bool isPalindrome(struct ListNode* head){
    struct ListNode* slow = head;
    struct ListNode* fast = head;
    struct ListNode* previous = NULL;
    while(fast != NULL && fast->next != NULL){
        slow = slow->next;
        fast = fast->next->next;
    }
    previous = slow;
    slow = slow->next;
    previous->next = NULL;
    while(slow != NULL){
        struct ListNode* temp = slow->next;
        slow->next = previous;
        previous = slow;
        slow = temp;
    }
    fast = head;
    slow = previous;
    while(slow != NULL){
        if(fast->val != slow->val){
            return false;
        }
        fast = fast->next;
        slow = slow->next;
    }
    return true;
}